src/backtrace/unwind.rs

```diff
@@ -0,0 +1,174 @@
+//! unwind target's program
+
+use anyhow::{bail, ensure, Context as _};
+use gimli::{
+    BaseAddresses, DebugFrame, LittleEndian, UninitializedUnwindContext, UnwindSection as _,
+};
+use probe_rs::{config::RamRegion, Core};
+
+use crate::{
+    cortexm,
+    registers::{self, Registers},
+    stacked::Stacked,
+    Outcome, VectorTable,
+};
+
+static MISSING_DEBUG_INFO: &str = "debug information is missing. Likely fixes:
+1. compile the Rust code with `debug = 1` or higher. This is configured in the `profile.{release,bench}` sections of Cargo.toml (`profile.{dev,test}` default to `debug = 2`)
+2. use a recent version of the `cortex-m` crates (e.g. cortex-m 0.6.3 or newer). Check versions in Cargo.lock
+3. if linking to C code, compile the C code with the `-g` flag";
+
+/// Virtually* unwinds the target's program
+/// \* destructors are not run
+// FIXME(?) this should be "infallible" and return as many frames as possible even in case of IO
+// errors
+pub(crate) fn target(
+    core: &mut Core,
+    debug_frame: &[u8],
+    vector_table: &VectorTable,
+    sp_ram_region: &Option<RamRegion>,
+) -> anyhow::Result<Output> {
+    let mut debug_frame = DebugFrame::new(debug_frame, LittleEndian);
+    debug_frame.set_address_size(cortexm::ADDRESS_SIZE);
+
+    let mut pc = core.read_core_reg(registers::PC)?;
+    let sp = core.read_core_reg(registers::SP)?;
+    let lr = core.read_core_reg(registers::LR)?;
+    let base_addresses = BaseAddresses::default();
+    let mut unwind_context = UninitializedUnwindContext::new();
+
+    let mut outcome = Outcome::Ok;
+    let mut registers = Registers::new(lr, sp, core);
+    let mut raw_frames = vec![];
+    let mut corrupted = true;
+
+    loop {
+        if cortexm::is_hard_fault(pc, vector_table) {
+            debug_assert!(
+                raw_frames.is_empty(),
+                "when present HardFault handler must be the first frame we unwind but wasn't"
+            );
```

Comment:
Is this a hard requirement? I'd prefer to make it a regular `assert!` if so. That would mean that HardFaul cannot call any functions though, right?

Reply:
`probe-run` puts a breakpoint on the HardFault handler (the trampoline in cortex-m-rt, not the user-overridable function) so the program is halted because any function can be called from the exception handler.

Reply:
Ah, right